jep/222/README.adoc

```diff
@@ -238,6 +247,56 @@ there are extra network round trips in the best case
 (to get the `X-Jenkins-JNLP-Port` header from Jenkins over HTTP, then to make a TCP connection);
 and in the worst case the TCP connection might hang rather than failing cleanly.
 
+=== More natural CLI framing
+
+Rather than reusing `PlainCLIProtocol` from the full-duplex HTTP transport,
+the WebSocket-based CLI endpoint could be designed to be friendlier to generic clients such as `websocat`.
+For example, `CLICommand.stdin` could be streamed from incoming frames,
+and `CLICommand.stdout` could be chunked into outgoing frames.
+Some features of the Jenkins CLI fit naturally into this model,
+such as the use of `Accept-Charset` and `Accept-Language` headers.
+
+However, several obstacles seemed to make this approach more trouble than it would be worth:
+
+* Unlike the SSH protocol, there is no simple way to enumerate a command name and arguments in the request:
+  you would need to use query parameters or HTTP headers in an awkward fashion.
+  (And ensuring that arguments containing spaces or other special characters are supported would complicate the scheme.)
+* Again unlike SSH, there is no standard way to differentiate `stdout` from `stderr`;
+  binary vs. text frames (respectively) could be used for this,
+  but generic clients are unlikely to honor the distinction.
+* Again unlike SSH, there is no standard way to represent an exit code:
+  an HTTP header is not an option in interactive mode (the exit code is determined after the upgrade response is sent),
+  so a special frame syntax would be needed.
+
+If and when there is a need for a protocol which can be used easily from a generic client,
+this could be implemented in a plugin.
+In fact the link:https://plugins.jenkins.io/cli-commander[CLI Commander plugin]
+already link:https://github.com/jenkinsci/cli-commander-plugin/blob/4831bf13711f5e382610980eedc815e4dd3f0c0d/src/main/java/org/jenkinsci/plugins/clicommander/Commander.java#L81-L138[does something similar]
+for the case of noninteractive commands
+(though in that case the primary use case is via a browser UI rather than something like `curl`).
+
+=== WebSocket mode by default for the CLI
+
+The new transport for the CLI could be made the default,
+perhaps a fallback to `-http` mode in case the server is too old or does not support WebSocket
```

Comment:
This fallback will be needed IMHO

Reply:
To be clear, this is in the **Reasoning** section—i.e., alternatives considered but not taken.